pyro/distributions/sine_skewed.py

```diff
@@ -0,0 +1,80 @@
+from math import pi
+
+import torch
+from torch.distributions import Uniform
+
+from pyro.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class SineSkewed(TorchDistribution):
+    """The Sine Skewed distribution is a distribution for breaking pointwise-symmetry on a base-distributions over
+    the d-dimensional torus.
+
+    This distribution requires a base distribution on the torus. The parameter skewness can be inferred using
+    :class:`~pyro.infer.HMC` or :class:`~pyro.infer.NUTS`. The following will produce a uniform prior
+    over skewness,::
+
+        def model(data, batch_shape, event_shape):
+            n = torch.prod(torch.tensor(event_shape), -1, dtype=torch.int)
+            skewness = torch.empty((*batch_shape, n)).view(-1, n)
+            tots = torch.zeros(batch_shape).view(-1)
+            for i in range(n):
+                skewness[..., i] = pyro.sample(f'skew{i}', Uniform(0., 1 - tots))
+                tots += skewness[..., i]
+            sign = pyro.sample('sign', Uniform(0., torch.ones(skewness.shape)).to_event(len(skewness.shape)))
+            skewness = torch.where(sign < .5, -skewness, skewness)
+
+            if (*batch_shape, *event_shape) == tuple():
+                skewness = skewness.reshape((*batch_shape, *event_shape))
+            else:
+                skewness = skewness.view(*batch_shape, *event_shape)
+
+    .. note:: The base-distribution must be over a arbitrary dim torus.
+
+    .. note:: ``skewness.abs().sum() <= 1.`` and ``(skewness.abs() <= 1).all()``.
+
+    ** References: **
+      1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+         Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param base_density: base density on the d-dimensional torus; event_shape must be [..., 2] where
+        ``prod(event_shape[:-1]) == d``.
+    :param skewness: skewness of the distribution; must have same shape as base_density.event_shape, all values
+        must be in [-1,1] and ``abs(skewness).sum() <= 1``.
```

Comment:
Should `skewness` also have batch_shape? In addition, why we need to support complicated `event_shape` here? Shouldn't it be enough to require `base_density.event_shape == (2,)`?

Reply:
> Should skewness also have batch_shape? 

Yes, you are right; I fixed this in code but forgot to update the docstring. 

> In addition, why we need to support complicated event_shape here? Shouldn't it be enough to require base_density.event_shape == (2,)?

This density works for any d-dimensional torus, so while the use cases I care/know are on the 1-torus, I figured it would be better to support the general case. What do you think? It's a minor change if you think it's appropriate.

Reply:
By d-dimensional torus, do you mean `SineSkewed(base_dist, skewness).to_event(d)` where `base_dist.event_shape == (2,)`?

Reply:
I see. Thanks for clarifying. I'll add more test cases and resolve this.